pkg/importer/data-processor.go

```diff
@@ -262,34 +274,56 @@ func (dp *DataProcessor) resize() (ProcessingPhase, error) {
 			err = errors.Wrap(err, "Unable to change permissions of target file")
 		}
 	}
+	if dp.preallocation && shouldPreallocate {
+		return ProcessingPhasePreallocate, nil
+	}
+	return ProcessingPhaseComplete, nil
+}
+
+func (dp *DataProcessor) preallocate() (ProcessingPhase, error) {
+	if !dp.preallocation {
+		return ProcessingPhaseComplete, nil
+	}
+
+	klog.V(3).Infoln("Preallocating")
+	// Preallocation is implemented as a copy from file to itself
+	destURL, _ := url.Parse(dp.dataFile)
+	err := qemuOperations.ConvertToRawStream(destURL, dp.dataFile, dp.preallocation)
+	if err != nil {
+		return ProcessingPhaseError, errors.Wrap(err, "Preallocation or resized image failed")
+	}
+
 	return ProcessingPhaseComplete, nil
 }
 
 // ResizeImage resizes the images to match the requested size. Sometimes provisioners misbehave and the available space
 // is not the same as the requested space. For those situations we compare the available space to the requested space and
 // use the smallest of the two values.
-func ResizeImage(dataFile, imageSize string, totalTargetSpace int64) error {
+func ResizeImage(dataFile, imageSize string, totalTargetSpace int64) (bool, error) {
 	dataFileURL, _ := url.Parse(dataFile)
 	info, err := qemuOperations.Info(dataFileURL)
 	if err != nil {
-		return err
+		return false, err
 	}
 	if imageSize != "" {
+		shouldPreallocate := true
 		currentImageSizeQuantity := resource.NewScaledQuantity(info.VirtualSize, 0)
 		newImageSizeQuantity := resource.MustParse(imageSize)
 		minSizeQuantity := util.MinQuantity(resource.NewScaledQuantity(totalTargetSpace, 0), &newImageSizeQuantity)
 		if minSizeQuantity.Cmp(newImageSizeQuantity) != 0 {
 			// Available destination space is smaller than the size we want to resize to
 			klog.Warningf("Available space less than requested size, resizing image to available space %s.\n", minSizeQuantity.String())
+			shouldPreallocate = false
 		}
 		if currentImageSizeQuantity.Cmp(minSizeQuantity) == 0 {
 			klog.V(1).Infof("No need to resize image. Requested size: %s, Image size: %d.\n", imageSize, info.VirtualSize)
-			return nil
+			return false, nil
 		}
 		klog.V(1).Infof("Expanding image size to: %s\n", minSizeQuantity.String())
-		return qemuOperations.Resize(dataFile, minSizeQuantity)
+		err := qemuOperations.Resize(dataFile, minSizeQuantity)
```

Comment:
Any reason you don't run the preallocation on the resize phase, since the resize phase is called regardless of the file was written by go client or qemu-img?

Reply:
I like clear separation of duties. If I do preallocation during that phase, it would make it a "resize-and-maybe-preallocate" phase.